### ISSUE TITLE
Refactor training loop into modular, injectable userApi functions

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#66]
## Summary

Break the monolithic training loop in `MnistExperiment.c` into a modular `training_loop/` module under `src/userApi/`. The module provides a layered hierarchy (run → epoch → batch) with **injectable function pointers** for `calculateGrads` and `inference`, enabling side-by-side comparison of different training strategies.

**Prerequisite:** Fix pre-existing bugs in #67 first.

## Motivation

The current `MnistExperiment.c` is a ~340-line monolith mixing model setup, training, evaluation, and logging. The training orchestration logic is not reusable. Additionally, `TrainingApi.c` has a misleadingly named `trainingEpoch` that actually does a single-sample forward+backward+optimizer step.

We want to:
1. Make training loop components reusable across experiments
2. Enable comparison of different `calculateGrads` / `trainingBatch` / `trainingEpoch` implementations (e.g., sequential vs accumulated gradients)
3. Support optional logging via callbacks (important for MCU targets where CSV is not available)
4. Keep evaluation separate (loss vs accuracy)

## Design Decisions (resolved)

| Decision | Resolution | Rationale |
|----------|-----------|-----------|
| `calculateGrads` + static helpers | Move to `CalculateGradsSequential.c` (option b) | Clean separation; helpers are implementation details of the sequential strategy |
| `trainingBatchDefault` return type | `float` (average loss) | Per-sample `trainingStats_t **` is wasteful — only loss is used in practice |
| Optimizer step+zero ownership | `trainingEpochDefault` (not batch) | Keeps batch pure (grad accumulation only); enables multi-batch accumulation strategies |
| `trainingRun` return type | `trainingRunResult_t` struct | `void` prevents programmatic convergence checks, especially on MCU with NULL callback |
| `TrainingApi.c` fate | **Remove entirely** | After moving `calculateGrads` out and removing `trainingEpoch`, only `calculateGradsBatched` + two free functions remain — all redundant with the new module |
| `trainingStats_t` + `freeTrainingStats` | Move to `TrainingLoopApi.h/.c` | `trainingStats_t` is the return type of `calculateGradsFn_t`, so it belongs in the module that defines the typedef |
| Individual args vs config struct | Individual arguments | Consistent with existing codebase style |
| Subdirectories per function category | Yes | Makes interchangeability visible in the filesystem; matches `userApi/layer/`, `userApi/optimizer/` patterns |
| Nullable epoch callback | Yes | MCU targets can't do CSV logging; NULL skips it |

## Directory Layout

```
src/userApi/training_loop/
├── CMakeLists.txt
├── TrainingLoopApi.c                    (trainingRun, evaluationEpoch, evaluationBatch, freeTrainingStats, etc.)
├── TrainingLoopApiInternal.h
├── calculate_grads/
│   ├── CMakeLists.txt
│   ├── CalculateGradsSequential.c       (calculateGrads + all static helpers, moved from TrainingApi.c)
│   └── include/
│       └── CalculateGradsSequential.h
├── training_batch/
│   ├── CMakeLists.txt
│   ├── TrainingBatchDefault.c
│   └── include/
│       └── TrainingBatchDefault.h
├── training_epoch/
│   ├── CMakeLists.txt
│   ├── TrainingEpochDefault.c
│   └── include/
│       └── TrainingEpochDefault.h
└── include/
    └── TrainingLoopApi.h                (types, function pointer typedefs, run + eval functions)
```

Each subdirectory (`calculate_grads/`, `training_batch/`, `training_epoch/`) holds interchangeable implementations of the same function signature. Adding a variant = drop a `.c` + `.h` + one `add_library` line.

## Types (in `TrainingLoopApi.h`)

```c
// Moved from TrainingApi.h
typedef struct trainingStats {
    tensor_t *output;
    float loss;
} trainingStats_t;

// Result of a full training run
typedef struct trainingRunResult {
    float finalTrainLoss;
    float finalEvalLoss;
} trainingRunResult_t;

// Injectable: forward + loss + backward for one sample
typedef trainingStats_t *(*calculateGradsFn_t)(
    layer_t **model, size_t modelSize, lossType_t lossType,
    tensor_t *input, tensor_t *label);

// Injectable: forward + loss for one sample
typedef inferenceStats_t *(*inferenceWithLossFn_t)(
    layer_t **model, size_t numberOfLayers, tensor_t *input,
    tensor_t *label, lossType_t lossType);

// Injectable: forward only for one sample
typedef tensor_t *(*inferenceFn_t)(
    layer_t **model, size_t numberOfLayers, tensor_t *input);

// Epoch logging callback (NULL to skip)
typedef void (*epochCallbackFn_t)(size_t epoch, float trainLoss, float evalLoss);
```

## Function Hierarchy

```
trainingRun → trainingRunResult_t
├── trainingEpochDefault → float    (per epoch — injectable, owns optimizer step+zero)
│   └── trainingBatchDefault → float  (per batch — injectable, pure grad accumulation)
│       └── calculateGradsFn           (per sample — injectable)
├── evaluationEpoch → float          (per epoch, loss)
│   └── evaluationBatch → float      (per batch, loss)
│       └── inferenceWithLossFn        (per sample — injectable)
└── epochCallbackFn                   (nullable)
```

Plus standalone accuracy evaluation:
```
evaluationEpochAccuracy → float
└── evaluationBatchAccuracy → size_t
    └── inferenceFn            (per sample — injectable)
```

### Batch level

```c
// training_batch/ — accumulates gradients over batch, returns average loss
// Does NOT call optimizer step+zero (that's the epoch's job)
float trainingBatchDefault(
    layer_t **model, size_t modelSize, lossType_t lossType,
    batch_t *batch, calculateGradsFn_t calculateGradsFn);

// TrainingLoopApi.c — returns average loss over batch
float evaluationBatch(
    layer_t **model, size_t modelSize, lossType_t lossType,
    batch_t *batch, inferenceWithLossFn_t inferenceFn);

// TrainingLoopApi.c — returns number of correct predictions
size_t evaluationBatchAccuracy(
    layer_t **model, size_t modelSize,
    batch_t *batch, size_t numClasses,
    inferenceFn_t inferenceFn);
```

### Epoch level

```c
// training_epoch/ — owns optimizer step+zero per batch, returns average loss
float trainingEpochDefault(
    layer_t **model, size_t modelSize, lossType_t lossType,
    dataLoader_t *dataLoader, optimizer_t *optimizer,
    calculateGradsFn_t calculateGradsFn);

// TrainingLoopApi.c — returns average eval loss
float evaluationEpoch(
    layer_t **model, size_t modelSize, lossType_t lossType,
    dataLoader_t *dataLoader,
    inferenceWithLossFn_t inferenceFn);

// TrainingLoopApi.c — returns accuracy as fraction
float evaluationEpochAccuracy(
    layer_t **model, size_t modelSize,
    dataLoader_t *dataLoader, size_t numClasses,
    inferenceFn_t inferenceFn);
```

### Run level

```c
// TrainingLoopApi.c — top-level orchestrator
trainingRunResult_t trainingRun(
    layer_t **model, size_t modelSize, lossType_t lossType,
    dataLoader_t *trainDataLoader, dataLoader_t *evalDataLoader,
    optimizer_t *optimizer, size_t numberOfEpochs,
    calculateGradsFn_t calculateGradsFn,
    inferenceWithLossFn_t inferenceFn,
    epochCallbackFn_t callback);
```

### Utilities

```c
// TrainingLoopApi.c — moved from TrainingApi.c
void freeTrainingStats(trainingStats_t *trainingStats);
```

### TrainingApi.c / TrainingApi.h — **REMOVE ENTIRELY**
- `calculateGrads` + static helpers (`initLayerOutputs`, `deInitLayerOutputs`, `initGradTensor`, `deInitGradTensor`, `initTrainingStats`) → move to `CalculateGradsSequential.c`
- `trainingStats_t` + `freeTrainingStats` → move to `TrainingLoopApi.h/.c`
- `trainingEpoch` → replaced by `TrainingEpochDefault`
- `calculateGradsBatched` → replaced by `trainingBatchDefault`
- `freeTrainingStatsBatched` → no longer needed (batch returns `float`)
- Update `src/userApi/CMakeLists.txt` to remove the `TrainingApi` library target
- Update any CMake targets that linked `TrainingApi` to link appropriate new targets instead

### InferenceApi.c / InferenceApi.h — **NO CHANGES**
- `inference` and `inferenceWithLoss` remain as-is; they serve as default injectables for the new function pointer typedefs

### MnistExperiment.c
- Simplify `main()` to: model setup → dataloader setup → `trainingRun(...)` → `evaluationEpochAccuracy(...)`
- Remove inline training/eval loops
- Remove `argmax` helper (moves to eval batch accuracy impl)
- Remove `writeCsvRow` (replaced by epoch callback)

## Implementation Order

1. Create `training_loop/` directory structure and CMakeLists
2. Define types and function pointer typedefs in `TrainingLoopApi.h`
3. Implement `freeTrainingStats` in `TrainingLoopApi.c`
4. Move `calculateGrads` + static helpers to `CalculateGradsSequential.c`
5. Implement `evaluationBatch`, `evaluationBatchAccuracy`, `evaluationEpoch`, `evaluationEpochAccuracy` in `TrainingLoopApi.c`
6. Implement `TrainingBatchDefault` (calls `calculateGradsFn` per sample, returns avg loss)
7. Implement `TrainingEpochDefault` (calls batch + optimizer step+zero per batch)
8. Implement `trainingRun` in `TrainingLoopApi.c`
9. Refactor `MnistExperiment.c` to use new API
10. Remove `TrainingApi.c`, `TrainingApi.h`, `TrainingApiInternal.h` and CMake target
11. Add unit tests

**Comment by LeoBuron** [comment]
Implemented in #72.
